### PR TITLE
Add Kernel Config Checks Platform Compatibility Tool

### PR DESCRIPTION
**Pull request by nimanch**
This PR Adds additional kernel configuration checks required by IoT Edge to run on a target device
1. Replaced the check to perform a setcap operation with a kernel configuration check
2. Added some additional todos

## Azure IoT Edge PR checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines and Best Practices
- [x] I have read the [contribution guidelines](https://github.com/azure/iotedge#contributing).
- [x] Title of the pull request is clear and informative.
- [x] Description of the pull request includes a concise summary of the enhancement or bug fix.

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.
- Description of the pull request includes 
	- [ ] concise summary of tests added/modified
	- [ ] local testing done.  

### Draft PRs
- Open the PR in `Draft` mode if it is:
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.

_Note: We use the kodiakhq bot to merge PRs once the necessary checks and approvals are in place. When it merges a PR, kodiakhq converts the PR title to the commit title, PR description to the commit description, and squashes all the commits in the PR to a single commit. The net effect is that entire PR becomes a single commit. Please follow the best practices mentioned [here](https://chris.beams.io/posts/git-commit/#:~:text=The%20seven%20rules%20of%20a%20great%20Git%20commit,what%20and%20why%20vs.%20how%20For%20example%3A%20) for the PR title and description_

### CODE REVIEW

File: platform-validation/scripts/compatibility.sh [hunk 1]
```diff
@@ -490,14 +531,31 @@ check_systemd(){
     fi
 }
 
+cat /etc/os-release
 get_architecture
 echo "Architecture:$ARCH"
 echo "OS Type:$OSTYPE"
 
-#TODO : Do we need to check in both host and container?
-check_net_cap_bind_host
```
Comment (ggjjj) [inline platform-validation/scripts/compatibility.sh:498]: did you remove this by mistake?

File: platform-validation/scripts/compatibility.sh [hunk 2]
```diff
@@ -490,14 +531,31 @@ check_systemd(){
...
-check_net_cap_bind_container
 check_cgroup_heirachy
+#Flags Required for setting elevated capabilities in a container.
+check_kernel_flags EXT4_FS_SECURITY
+
+#Todo : Only check if docker engine is not present?
+#Check for Required Container Engine Flags if docker is not present
+check_kernel_flags \
+	NAMESPACES NET_NS PID_NS IPC_NS UTS_NS \
```
Comment (ggjjj) [inline platform-validation/scripts/compatibility.sh:546]: Can you have a comment for each flag and what the flag corresponds to

File: platform-validation/scripts/compatibility.sh [hunk 3]
```diff
@@ -357,11 +356,53 @@ perform_cleanup(){
...
+check_kernel_flags(){
+    EXIT_CODE=0
+    check_kernel_file
+    if [ $EXIT_CODE != 0 ]; then 
+        wrap_fail "check_kernel_flags"
+        return
+    fi
+    EXIT_CODE=0
+    wrap_debug "Reading Kernel Config from $CONFIG"
+    for flag in "$@"; do
```
Comment (ggjjj) [inline platform-validation/scripts/compatibility.sh:399]: $@ corresponds to all the command line arguments right? The plan was to pass the first argument as "config path" like ./compatibility.sh "/path1 /path2 /path3" or ./compatibility.sh "/path1"  "/path2"  "/path3"

File: platform-validation/scripts/compatibility.sh [hunk 4]
```diff
@@ -490,14 +531,31 @@ check_systemd(){
...
 get_architecture
 echo "Architecture:$ARCH"
 echo "OS Type:$OSTYPE"
 
-#TODO : Do we need to check in both host and container?
-check_net_cap_bind_host
-check_net_cap_bind_container
 check_cgroup_heirachy
+#Flags Required for setting elevated capabilities in a container.
+check_kernel_flags EXT4_FS_SECURITY
```
Comment (ggjjj) [inline platform-validation/scripts/compatibility.sh:541]: we could let Cx know if the elevation is not met as a warning and launch with right priveleges?

File: platform-validation/scripts/compatibility.sh (hunk 3 above)
Comment (nimanch) [reply platform-validation/scripts/compatibility.sh:399]: No, $@ refers to the arguments passed to the functions themselves

File: platform-validation/scripts/compatibility.sh (hunk 4 above)
Comment (nimanch) [reply platform-validation/scripts/compatibility.sh:580]: No, The Customer would need to rebuild the kernel.

File: platform-validation/scripts/compatibility.sh (hunk 2 above)
Comment (nimanch) [reply platform-validation/scripts/compatibility.sh]: I added a general comment on why these checks are needed and provided a link to description on the config flags. I don't think we are at a stage where we know exactly how each configuration flag prevents us from running iot edge. We just have an idea that these flags are required for running a container engine which is a requirement for IoT Edge and hence I am hesitant to include any description in the code itself

Comment (ggjjj) [reply platform-validation/scripts/compatibility.sh]: where is the comment?

**Review by ggjjj — APPROVED** [review]

**Pull request merged**